game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):
+        self._board = board
+        self._player_next = player_next
+        self._connect_number = connect_number
+        self._number_of_moves = number_of_moves
+        self._moves_list = moves_list
+        # self._last_move = last_move
+        # self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = deepcopy(self._player_next)
+        connect_number = deepcopy(self._connect_number)
+        number_of_moves = deepcopy(self._number_of_moves)
+        moves_list = deepcopy(self._moves_list)
+        # previous_game_state = deepcopy(self)
+        return GameState(board, player_next, connect_number, number_of_moves, moves_list)
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def get_last_move(self):
+        return self._moves_list[-1]
+    
+    def get_connect_number(self):
+        return self._connect_number
+
+    def get_number_of_moves(self):
+        return self._number_of_moves
+    
+    def make_move(self, move, player_next = True):
+               
+        self._board[move[0]][move[1]] = self._player_next
+        self._player_next = - self._player_next
+        self._moves_list.append(move)
+        self._number_of_moves = self._number_of_moves + 1
+
+        # return game_state
+        
+    def unmake_move(self):

```

Comment:
I'd prefer to see safety built-in to this method, rather than relying on callers/users to do the right thing. In your `main()` function somewhere you have code to disable this getting called in a bad spot:
```
if game_state.get_number_of_moves() == 0:
    break
```

I'd recommend still keeping the above code (or similar), but also have this method have clearer and more-understandable behavior when called in a game-start situation. I'd probably start the function with an if-statement and then throw an exception with a clear message of what's wrong (we can talk about custom exception-classes at some point, but just a normal `Exception` isn't bad).

These are known as guard-clauses and some reasons for using them are here: https://maximegel.medium.com/what-are-guard-clauses-and-how-to-use-them-350c8f1b6fd2

Reply:
Also note that I'd make safety-code like this a higher priority now that this class has moved from immutable to mutable (as you mention in your email, for performance reasons, but there could be many other reasons to do this also). The codebase as a whole is still small right now, but complex behavioral interaction like this quickly multiply and can lead to gnarly, hard-to-understand bugs.